cli/connect/history.go

```diff
@@ -0,0 +1,133 @@
+package connect
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/tarantool/tt/cli/util"
+)
+
+// commandHistory stores console command history.
+type commandHistory struct {
+	filepath    string
+	commands    []string
+	timestamps  []int64
+	maxCommands int
+}
+
+// parseHistoryCells extracts timestamps and commands from history lines.
+// If format doesn't contain timestamps, it sets the current timestamp
+// for each command.
+func parseHistoryCells(
+	lines []string) (commands []string, timestamps []int64) {
+	regex := regexp.MustCompile(`^#\d+$`)
+
+	// startPos is the first position of a timestamp.
+	startPos := -1
+	for i, line := range lines {
+		if regex.MatchString(line) {
+			startPos = i
+			break
+		}
+	}
+	timestamps = make([]int64, 0)
+	if startPos == -1 {
+		// Read one line per command.
+		// Set the current timestamp for each command.
+		commands = lines
+		timestamp := time.Now().Unix()
+		for range lines {
+			timestamps = append(timestamps, timestamp)
+		}
+		return
+	}
+
+	commands = make([]string, 0)
+	for startPos < len(lines) {
+		j := startPos + 1
+
+		// Move pointer to the next timestamp.
+		for j < len(lines) && !regex.MatchString(lines[j]) {
+			j++
+		}
+
+		// Extract the current timestamp.
+		timestamp, err := strconv.Atoi(lines[startPos][1:])
+
+		if j != startPos+1 && err == nil {
+			timestamps = append(timestamps,
+				int64(timestamp))
+			commands = append(commands,
+				strings.Join(lines[startPos+1:j], "\n"))
+		}
+		startPos = j
+	}
+
+	return
+}
+
+// load loads console history from the history file.
+func (history *commandHistory) load() error {
+	if history == nil {
+		return nil
+	}
+
+	rawLines, err := util.GetLastNLines(history.filepath, history.maxCommands)
+	if err != nil {
+		return err
+	}
+
+	history.commands, history.timestamps = parseHistoryCells(rawLines)
+	return nil
+}
+
+// appendCommand appends new command to the history.
+func (history *commandHistory) appendCommand(command string) {
+	if history == nil {
+		return
+	}
+
+	history.commands = append(history.commands, command)
+	history.timestamps = append(history.timestamps, time.Now().Unix())
+	if len(history.commands) > history.maxCommands {
+		history.commands = history.commands[1:]
+		history.timestamps = history.timestamps[1:]
+	}
+}
+
+// writeToFile writes console history to the file.
+func (history *commandHistory) writeToFile() error {
+	if history == nil {
+		return nil
+	}
+
+	historyContent := bytes.Buffer{}
+	for i, command := range history.commands {
+		historyContent.WriteString(fmt.Sprintf("#%d\n%s\n", history.timestamps[i], command))
+	}
+	if err := os.WriteFile(history.filepath, historyContent.Bytes(), 0644); err != nil {
+		return fmt.Errorf("failed to write to history file: %s", err)
+	}
+
+	return nil
+}
```

Comment:
Let's consider the following scenario.

Console 1:

```sh
$ tt connect uri1
uri1> bear
uri1> wolf
uri1> hippo
uri1>
```

Console 2:

```sh
$ tt connect uri2
uri2> banana
uri2> orange
uri2> lemon
uri2>
```

Swich to console 1:

```sh
uri1> Ctrl+D
```

Switch to console 2:

```sh
uri2> Ctrl+D
```

Will animal and fruit commands saved both?

Reply:
No, the history of only one connection will be saved. Tarantool acts the same way.

Reply:
Tarantool drops the history to the file after each command. Nope, everything is OK there with several simultaneously opened consoles.

Reply:
A little confusion occured here.

First, the scenario works differently if we open two tarantool/tt sessions first and if we type the commands in the first session and then open the second one. The reason is that the history is loaded at the process start.

From Albert's words -- the file is written per command, just like in tarantool. So the behavior should be the same.